src/convnets/resnets.jl

```diff
@@ -0,0 +1,543 @@
+# resnet.jl
+## It is recommended to check out the user's guide for more information
+## regarding the use of these functions.
+
+### ResNet blocks
+## These functions return a block to be used inside of a ResNet model.
+## The individual arguments are explained in the documentation of the functions.
+## Note that for these blocks to be used by the `_make_blocks` function, they must define
+## a dispatch `expansion(::typeof(fn))` that returns the expansion factor of the block 
+## (i.e. the multiplicative factor by which the number of channels in the input is increased).
+## The `_make_blocks` function will then call the `expansion` function to determine the
+## expansion factor of each block and use this to construct the stages of the model.
+
+"""
+    basicblock(inplanes, planes; stride = 1, downsample = identity, cardinality = 1,
+               base_width = 64, reduce_first = 1, dilation = 1,
+               first_dilation = nothing, activation = relu, norm_layer = BatchNorm,
+               drop_block = identity, drop_path = identity,
+               attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
+
+Creates a basic ResNet block.
+
+# Arguments
+
+  - `inplanes`: number of input feature maps
+  - `planes`: number of feature maps for the block
+  - `stride`: the stride of the block
+  - `downsample`: the downsampling function to use
+  - `cardinality`: redundant, kept for compatibility with `bottleneck`.
+  - `base_width`: redundant, kept for compatibility with `bottleneck`.
+  - `reduce_first`: the reduction factor that the input feature maps are reduced by before the first
+    convolution.
+  - `dilation`: the dilation of the second convolution.
+  - `first_dilation`: the dilation of the first convolution.
+  - `activation`: the activation function to use.
+  - `norm_layer`: the normalization layer to use.
+  - `drop_block`: the drop block layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `drop_path`: the drop path layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `attn_fn`: the attention function to use. See [`squeeze_excite`](#) for an example.
+  - `attn_args`: a NamedTuple that contains none, some or all of the arguments to be passed to the
+    attention function.
+"""
+function basicblock(inplanes, planes; stride = 1, downsample = identity, cardinality = 1,
+                    base_width = 64, reduce_first = 1, dilation = 1,
+                    first_dilation = dilation, activation = relu, norm_layer = BatchNorm,
+                    drop_block = identity, drop_path = identity,
+                    attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
```

Comment:
I didn't quite get you here. `attn_fn` will in cases like `SE-ResNet` be `squeeze_excite`. `attn_args` is to pass in kwargs to that function

Reply:
I mean something like `attn_fn = planes -> squeeze_excite(planes; reduction = 64)`, i.e. closing over the keyword arguments. I think having fewer keywords that need to work in conjunction is cleaner, and this way the user says exactly what they want.

With `block_args` it was a bit different because there are so many customizations that a closure might be too cumbersome to write despite being simpler to understand. Here I don't think that's the case.

Reply:
That does make sense, but more closures will mean more cryptic `var#x#y` displays inside of the `Chain`. If that's okay here I don't mind this change but this is something that caught my eye in a lot of places because debugging is a nightmare when you don't have the function name